### ISSUE TITLE
Instructions on how to update ADS won't update the Ivadomed package

### ISSUE THREAD

**Issue by Stoyan-I-A** [axondeepseg/axondeepseg#623]
While testing #620, I encountered an error that was previously fixed on the Ivadomed side. This was because I didn't update the Ivadomed package in my environment.

Currently, in our docs, it says that we can update ADS by using these commands:
```
cd axondeepseg
git pull
pip install -e .
```

However, those instructions won't update the Ivadomed package. The user would need to do an additional command:
`pip install ivadomed==<The latest version supported by ADS>`.

Question: Will ADS always support the latest Ivadomed version? In this case, we would just need to add `pip install ivadomed` to the instructions. But if we insist on installing a fixed version, the user would have to check in the environment.yml file which version to install, which might be a bit cumbersome for the user.

**Comment by mathieuboudreau** [comment]
Good catch.

Not only will it not update the IVADOMED package, but it won't update any new/updated requirements in the conda file. So I think the problem is larger than just giving instructions on install IVADOMED.

We'll need to change the update ADS installation to reinstall the entire environment.yml using CONDA inside the venv.

**Comment by hermancollin** [comment]
I did some tests and it is indeed quite impossible to update all the pip packages of the conda environment automatically. There is an experimental feature in conda 4.6 for [improved pip interoperability](https://docs.conda.io/projects/conda/en/latest/user-guide/configuration/pip-interoperability.html) but the simplest and safest way is still to reinstall the whole environment, unfortunately.

**Comment by hermancollin** [comment]
Resolution of #617 by bumping ivadomed to 2.9.5 raises again the problem mentioned here.

It got me thinking that for the ivadomed package specifically, we should automate the update process since the ADS's development will now be closely related to ivadomed's from now on. I messed around with it a bit and we can run post-install commands. This is a modified part of `setup.py`:

```python
def update_ivadomed():
    """Ivadomed is a pip package: needs manual update"""
    with open('environment.yml', 'r') as env_file:
        lines = env_file.readlines()
        for l in lines:
            if "ivadomed==" in l:
                version = l[16:-1]
                subprocess.run(["pip", "install", f"ivadomed=={version}"])
                break

class PostDevelopCommand(develop):
    """Post-installation for installation mode."""
    def run(self):

        develop.run(self)
        subprocess.check_call("download_models")
        subprocess.check_call("download_tests")
        atexit.register(update_ivadomed)

```

Of course all the other pip packages will still need to be manually updated but it would be less of an issue. Should I open a PR for this?

**Comment by mariehbourget** [comment]
Quick update: as I mentioned in last week's meeting, there is a PR in ivadomed ([1000](https://github.com/ivadomed/ivadomed/pull/1000)) that I think will help address these issues, and another related PR for documentation ([1125](https://github.com/ivadomed/ivadomed/pull/1125)).

We discussed it brieftly in ivadomed meeting today and @dyt811 will put together a meeting with ivadomed, SCT and ADS dev to discuss further.

**Comment by kousu** [comment]
You need to be wary about running code in setup.py. PyPA is trying really hard (but advertising really, really poorly) that they are deprecating setup.py; they want people to use [setup.cfg](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html?highlight=declarative#configuring-setuptools-using-setup-cfg-files) instead, which is a declarative file, meaning it cannot mold itself to its environment and there's only ever one correct output from it.

Particularly, running `pip` inside of setup.py is not going to work reliably. The procedural code in setup.py is stripped out when built into a .whl file. It might work, for a while, if you force people to install ADS from source.

Does `pip install --upgrade -e .` solve the original issue?

**Comment by hermancollin** [comment]
> Particularly, running `pip` inside of setup.py is not going to work reliably. The procedural code in setup.py is stripped out when built into a .whl file. It might work, for a while, if you force people to install ADS from source, but it's not really welcomed by the PyPA people.
> 

@kousu yeah... I read your discussion with PyPA devs and their opinion about it. Very informative btw. I agree that this is not a long-term solution and we might want to avoid merging PR #631 in its current state.

I would be curious to see how other big "pro" python packages deal with these kinds of issues. We can't be the first group to encounter this torch dependency can of worms.

> Does `pip install --upgrade -e .` solve the original issue?

Hmm it doesn't work on my side.

I mean in the end, if we can't find an appropriate solution for this, we could always ask the user to reinstall the whole environment from scratch like previously suggested here (also, from what I understood, this is how sct deals with this)

**Comment by mathieuboudreau** [comment]
Stale - no longer using ivadomed
